gym/envs/doom/doom_env.py

```diff
@@ -1,42 +1,117 @@
-    metadata = {'render.modes': ['human', 'rgb_array']}
+    metadata = {'render.modes': ['human', 'rgb_array'], 'video.frames_per_second': 35}
 
     def __init__(self):
         utils.EzPickle.__init__(self)
+        self.game = DoomGame()
+        self.mode = 'fast'                          # 'human', 'fast' or 'normal'
+        self.no_render = False                      # To disable double rendering in human mode
+        self.config = ''                            # Configuration file
+        self.scenario = ''                          # Scenario file
+        self.map = ''                               # Map
+        self.difficulty = 5                         # Difficulty (1 = Easy, 10 = Impossible)
+        self.viewer = None
+        self.screen_height = 480
+        self.screen_width = 640
+        self.action_space = spaces.HighLow(
+            np.matrix([[0, 1, 0]] * 38 + [[-10, 10, 0]] * 2 + [[-100, 100, 0]] * 3, dtype=np.int8))
+        self.observation_space = spaces.Box(low=0, high=255, shape=(self.screen_height, self.screen_width, 3))
+        self.allowed_actions = list(range(43))
 
     def _step(self, action):
-        # action is a np array but DoomGame.make_action expects a list of ints
-        list_action = [int(x) for x in action]
+        if 43 != len(action):
+            logger.warn('Doom action list must contain 43 items. Padding missing items with 0')
+            old_action = action
+            action = [0] * 43
+            for i in range(len(old_action)):
+                action[i] = old_action[i]
+        # action is a list of numbers but DoomGame.make_action expects a list of ints
+        if len(self.allowed_actions) > 0:
+            list_action = [int(action[x]) for x in self.allowed_actions]
+        else:
+            list_action = [int(x) for x in action]
         try:
-            state = self.game.get_state()
             reward = self.game.make_action(list_action)
+            state = self.game.get_state()
+            info = self._get_game_variables(state.game_variables, self.game.get_total_reward())
+
             if self.game.is_episode_finished():
                 is_finished = True
+                return np.zeros(shape=self.observation_space.shape, dtype=np.uint8), reward, is_finished, info
             else:
                 is_finished = False
-            return state.image_buffer.copy(), reward, is_finished, {}
+                return state.image_buffer.copy(), reward, is_finished, info
 
         except doom_py.vizdoom.ViZDoomIsNotRunningException:
-            return [], 0, True, {}
+            return np.zeros(shape=self.observation_space.shape, dtype=np.uint8), 0, True, {}
 
     def _reset(self):
-        self.game.new_episode()
-        return self.game.get_state().image_buffer.copy()
+        self._closed = False
+        package_directory = os.path.dirname(os.path.abspath(__file__))
+
+        # Common settings
+        self.loader = Loader()
+        self.game = DoomGame()
```

Comment:
I think we should move this initialization to a helper function and call it during __init__. When training an agent reset() gets called at the end of every trajectory, so it should be a fast operation in the common case. 

If we need the ability to change maps (e.g. for transfer learning) we should trigger it as a result of taking an action.